# The main entry point of your workflow.
# After configuring, running snakemake -n in a clone of this repository should
# successfully execute a dry-run of the workflow.
include: "rules/common.smk"
include: "rules/assembly.smk"
include: "rules/plasmids.smk"
include: "rules/crispr.smk"

report: "report/workflow.rst"

# this container defines the underlying OS for each job when using the workflow
# with --use-conda --use-singularity
container: "docker://continuumio/miniconda3:4.8.2"

scapp_output = expand("results/scapp/{assembly}/{assembly}.confident_cycs.fasta",
               assembly = assemblies.keys())
recycler_output = expand("results/recycler/{assembly}/{assembly}.cycs.fasta",
               assembly = assemblies.keys())
minced_output = expand("results/minced/{assembly}/{assembly}.crisprs.{s}.gz",
                         assembly = assemblies.keys(), s = ["txt", "gff"])
megahit_output = expand("results/assembly/{assembly}/final.contigs.fastg",
               assembly = assemblies.keys())

rule all:
    input:
        scapp = scapp_output,
        recycler = recycler_output,
        minced = minced_output

rule assemble:
    input:
        megahit = megahit_output

rule plasmids:
    input:
        scapp = scapp_output,
        recycler = recycler_output,

rule crisprs:
    input:
        minced = minced_output